pci/src/vfio.rs

```diff
@@ -376,18 +376,17 @@ impl VfioCommon {
             let region_size: u64;
             let bar_addr: GuestAddress;
 
-            let restored_bar_addr = if let Some(resources) = &resources {
-                match resources
-                    .get(bars.len())
-                    .ok_or(PciDeviceError::MissingResource)?
-                {
-                    Resource::MmioAddressRange { base, .. } => Some(GuestAddress(*base)),
-                    Resource::PioAddressRange { base, .. } => Some(GuestAddress(*base as u64)),
-                    _ => return Err(PciDeviceError::InvalidResourceType),
+            let mut restored_bar_addr = None;
+            if let Some(resources) = &resources {
+                for resource in resources {
+                    if let Resource::PciBar { index, base, .. } = resource {
+                        if *index == bar_id as usize {
+                            restored_bar_addr = Some(GuestAddress(*base));
+                            break;
+                        }
+                    }
```

Comment:
I assume we are expecting to restore all BARs when the input `resouces` is `Some`. So we should return an error if `restored_bar_addr` is `None` here  (e.g. missing a matching `resource`), right?

Reply:
No because we're going through every possible bar index, meaning we might not have a BAR to restore for each index.
But I want to improve the logic here by simplifying the restore path as it would rely entirely on the list of resources, rather than iterating through every possible bar index.

Reply:
I've added an extra commit, PTAL